switchboard/manager.py

```diff
@@ -214,9 +214,20 @@ def as_request(self, user=None, ip_address=None):
         return MockRequest(user, ip_address)
 
     def version_switch(self, switch):
-        if hasattr(switch, 'save_version'):
-            user = self.context.get('user')
-            switch.save_version(username=user.username if user else '')
+        # Try to get the username from both User objects and user dicts.
+        try:
+            user = self.context.get('user', {})
+            if hasattr(user, 'username'):
+                username = user.username
+            else:
+                username = user.get('username', '')
+        except:
```

Comment:
These try/excepts are more for "no matter what, don't cause other things to break", so they are intentionally broad and catch-all. Saving all changes to a switch so that we can audit a switch history is nice and all, but if something (anything) blows up it shouldn't prevent the changes from being saved.